File: app/extract.py
"""Turn raw Classroom posts into action items.

Uses the Claude API when ANTHROPIC_API_KEY is set; otherwise falls back to a
conservative heuristic that only creates low-confidence items when it sees a
clear signal (a date, "tomorrow", an amount, or an action keyword).
"""
import json
import os
import re
from datetime import date, datetime, timedelta

from dateutil import parser as dateparse

from . import db

ACTION_WORDS = re.compile(
    r"\b(submit|bring|send|pay|due|last date|deadline|test|exam|fill|wear|"
    r"attend|meeting|holiday|fee|fees|remind|prepare|complete|carry)\b", re.I)
DATE_PAT = re.compile(r"\b(\d{1,2})[./-](\d{1,2})[./-](\d{2,4})\b")
DATE_WORD_PAT = re.compile(  # e.g. "20-Jul-2026", "3 July 2026"
    r"\b(\d{1,2})[ \-]([A-Za-z]{3,9})[ \-,]+(\d{4})\b")
HOLIDAY_PAT = re.compile(r"\bholiday|school (will remain |remains )?closed\b", re.I)
AMOUNT_PAT = re.compile(r"(?:₹|rs\.?|inr)\s?([\d,]+)", re.I)
BULLET_PAT = re.compile(r"^\s*(?:[•\-\*•]|\d+[.)])\s+(.{3,120})$", re.M)

EXTRACT_MODEL = os.environ.get("EXTRACT_MODEL", "claude-sonnet-5")

WEEKDAY_KEYS = ["mon", "tue", "wed", "thu", "fri", "sat", "sun"]
DEFAULT_SCHOOL_DAYS = ["mon", "tue", "wed", "thu", "fri"]


def school_days_for(child_row):
    """A child's school days, derived from their timetable keys (a 'sat' key
    means Saturday classes). Falls back to Mon-Fri."""
    try:
        grid = json.loads(child_row.get("timetable_json") or "{}")
    except ValueError:
        grid = {}
    days = [d for d in WEEKDAY_KEYS if d in grid]
    return days or DEFAULT_SCHOOL_DAYS


def next_school_day(after, school_days):
    """The next school day strictly after `after` — so 'tomorrow' written on a
    Friday resolves to Monday (or Saturday, if the school has Saturday classes)."""
    for i in range(1, 8):
        d = after + timedelta(days=i)
        if WEEKDAY_KEYS[d.weekday()] in school_days:
            return d
    return after + timedelta(days=1)

PROMPT = """You are the extraction engine of ParentOps, an app that turns school
Google Classroom posts into action items for parents in India.

Post metadata:
- Course/subject: {course}
- Posted on: {posted} (use this to resolve relative dates like "tomorrow")
- Kind: {kind}
- This child's school days: {school_days}. Resolve "tomorrow"/"next class" to
  the NEXT SCHOOL DAY — "tomorrow" written on a Friday means Monday, unless
  Saturday is listed as a school day.

Post text:
---
{body}
---

Extract parent-facing action items. Return ONLY a JSON array (no prose, no code
fences). Each element:
{{
  "title": "short imperative summary, max 80 chars",
  "detail": "one-line context for the parent, or null",
  "due_date": "YYYY-MM-DD or null",
  "date_confidence": "explicit" | "inferred" | "none",
  "amount": "e.g. ₹1,250, or null",
  "category": "holiday" | "fee" | "test" | "event" | "task",
  "checklist": ["specific sub-task", ...]   // [] if none
}}

Rules:
- Purely informational posts (new study material, no action needed) -> []
- EXCEPTION: holiday / school-closed announcements DO produce an item with
  category "holiday" and the holiday date as due_date.
- "tomorrow"/"next Friday" etc.: resolve against the posted date and mark
  date_confidence "inferred", never "explicit". Note the inference in detail
  (e.g. 'teacher wrote "tomorrow"').
- Never invent deadlines, amounts, or checklist entries not present in the text.
- Checklist entries must be concrete physical actions (photos to send, pages to
  fill, items to bring), not restatements of the title."""


def _find_due(body, posted, school_days):
    m = DATE_PAT.search(body)
    if m:
        d, mo, y = int(m[1]), int(m[2]), int(m[3])
        y = y + 2000 if y < 100 else y
        try:
            return date(y, mo, d).isoformat(), "explicit"
        except ValueError:
            pass
    m = DATE_WORD_PAT.search(body)
    if m:
        try:  # "20-Jul-2026" / "3 July 2026"
            return dateparse.parse(" ".join(m.groups()), dayfirst=True).date().isoformat(), "explicit"
        except (ValueError, OverflowError):
            pass
    if re.search(r"\btomorrow\b", body, re.I) and posted:
        return next_school_day(posted, school_days).isoformat(), "inferred"
    return None, "none"


def _heuristic(post, school_days):
    body = f"{post.get('title') or ''}\n{post.get('body') or ''}"
    is_holiday = bool(HOLIDAY_PAT.search(body))
    if not (ACTION_WORDS.search(body) or is_holiday):
        return []
    posted = _posted_date(post)
    due, confidence = _find_due(body, posted, school_days)
    amount = None
    am = AMOUNT_PAT.search(body)
    if am:
        amount = f"₹{am[1]}"
    checklist = [{"text": t.strip(), "done": False} for t in BULLET_PAT.findall(body)][:8]
    if not (due or amount or checklist or is_holiday):
        return []
    if is_holiday:
        category = "holiday"
    elif amount:
        category = "fee"
    elif re.search(r"\btest|exam\b", body, re.I):
        category = "test"
    else:
        category = "task"
    return [{
        "title": (post.get("title") or "School notice")[:80],
        "detail": "Auto-detected without AI — open the original post to verify.",
        "due_date": due, "date_confidence": confidence,
        "amount": amount, "category": category, "checklist": checklist,
    }]


def _posted_date(post):
    try:
        return dateparse.parse(post.get("posted_at")).date()
    except (TypeError, ValueError):
        return None


def _claude(post, school_days):
    import anthropic
    client = anthropic.Anthropic()
    msg = client.messages.create(
        model=EXTRACT_MODEL, max_tokens=1500,
        messages=[{"role": "user", "content": PROMPT.format(
            course=post.get("course_name") or "unknown",
            posted=post.get("posted_at") or "unknown",
            kind=post.get("kind"),
            school_days=", ".join(school_days),
            body=(post.get("title") or "") + "\n" + (post.get("body") or "")[:6000],
        )}])
    text = msg.content[0].text.strip()
    text = re.sub(r"^```(?:json)?|```$", "", text, flags=re.M).strip()
    items = json.loads(text)
    for it in items:
        it["checklist"] = [{"text": t, "done": False} for t in it.get("checklist") or []]
    return items


def extract_post(post, school_days=None):
    """Extraction for one post -> list of item dicts (schema of PROMPT)."""
    school_days = school_days or DEFAULT_SCHOOL_DAYS
    if os.environ.get("ANTHROPIC_API_KEY"):
        try:
            return _claude(post, school_days)
        except Exception as e:  # fall back rather than lose the post
            print(f"[extract] Claude failed for post {post['id']}: {e}")
    return _heuristic(post, school_days)


def run_pending():
    """Extract all posts not yet processed. Returns number of items created."""
    created = 0
    days_by_child = {ch["id"]: school_days_for(ch) for ch in db.children()}
    for post in db.unextracted_posts():
        for it in extract_post(post, days_by_child.get(post["child_id"])):
            db.add_item(
                post["child_id"], it["title"][:120], post_id=post["id"],
                detail=it.get("detail"), due_date=it.get("due_date"),
                amount=it.get("amount"), checklist=it.get("checklist"),
                category=it.get("category"))
            created += 1
        db.mark_extracted(post["id"])
    return created

File: android/app/src/main/java/com/parentops/app/Extract.kt
package com.parentops.app

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.TextStyle
import java.util.Locale

/**
 * On-device extraction of action items from Classroom posts — a Kotlin port
 * of the server heuristic: explicit dates ("03.07.2026", "20-Jul-2026"),
 * "tomorrow" resolved to the next school day (Friday -> Monday unless the
 * child's timetable has Saturday classes), amounts, bullets, categories.
 */
object Extract {

    private val ACTION = Regex(
        "\\b(submit|bring|send|pay|due|last date|deadline|test|exam|fill|wear|" +
        "attend|meeting|holiday|fee|fees|remind|prepare|complete|carry)\\b",
        RegexOption.IGNORE_CASE)
    private val DATE_NUM = Regex("\\b(\\d{1,2})[./-](\\d{1,2})[./-](\\d{2,4})\\b")
    private val DATE_WORD = Regex("\\b(\\d{1,2})[ \\-]([A-Za-z]{3,9})[ \\-,]+(\\d{4})\\b")
    private val HOLIDAY = Regex("\\bholiday|school (will remain |remains )?closed\\b", RegexOption.IGNORE_CASE)
    private val AMOUNT = Regex("(?:₹|rs\\.?|inr)\\s?([\\d,]+)", RegexOption.IGNORE_CASE)
    private val BULLET = Regex("^\\s*(?:[•\\-*]|\\d+[.)])\\s+(.{3,120})$", RegexOption.MULTILINE)
    private val TOMORROW = Regex("\\btomorrow\\b", RegexOption.IGNORE_CASE)

    private val MONTHS = (1..12).associateBy {
        java.time.Month.of(it).getDisplayName(TextStyle.SHORT, Locale.ENGLISH).lowercase()
    }

    val WEEKDAY_KEYS = listOf("mon", "tue", "wed", "thu", "fri", "sat", "sun")
    private val DEFAULT_SCHOOL_DAYS = listOf("mon", "tue", "wed", "thu", "fri")
    private val GREETING = Regex(
        "^(dear|respected|hello|hi|greetings)\\b|^(regards|thanks|thank you)\\b",
        RegexOption.IGNORE_CASE)

    /** First meaningful line — skips "Dear Parents," style greetings. */
    fun headline(text: String): String {
        for (raw in text.lineSequence()) {
            val line = raw.trim()
            if (line.length < 4 || GREETING.containsMatchIn(line)) continue
            return line.take(120)
        }
        return text.lineSequence().firstOrNull { it.isNotBlank() }?.take(120) ?: "Note"
    }

    /** Short body preview for the card, greeting lines removed. */
    fun snippet(text: String, title: String): String? {
        val body = text.lineSequence()
            .map { it.trim() }
            .filter { it.isNotEmpty() && !GREETING.containsMatchIn(it) && it != title.trim() }
            .joinToString(" ")
        return body.take(200).ifBlank { null }
    }

    fun schoolDays(child: Child): List<String> {
        val days = WEEKDAY_KEYS.filter { child.timetable.containsKey(it) }
        return days.ifEmpty { DEFAULT_SCHOOL_DAYS }
    }

    fun nextSchoolDay(after: LocalDate, schoolDays: List<String>): LocalDate {
        for (i in 1..7) {
            val d = after.plusDays(i.toLong())
            if (WEEKDAY_KEYS[d.dayOfWeek.value - 1] in schoolDays) return d
        }
        return after.plusDays(1)
    }

    private fun postedDate(post: Post): LocalDate? = try {
        Instant.parse(post.postedAt).atZone(ZoneId.systemDefault()).toLocalDate()
    } catch (e: Exception) {
        null
    }

    private fun findDue(body: String, posted: LocalDate?, schoolDays: List<String>): String? {
        DATE_NUM.find(body)?.let { m ->
            val (d, mo, yRaw) = m.destructured
            val y = yRaw.toInt().let { if (it < 100) it + 2000 else it }
            try {
                return LocalDate.of(y, mo.toInt(), d.toInt()).toString()
            } catch (e: Exception) { /* not a real date */ }
        }
        DATE_WORD.find(body)?.let { m ->
            val (d, monName, y) = m.destructured
            MONTHS[monName.lowercase().take(3)]?.let { mo ->
                try {
                    return LocalDate.of(y.toInt(), mo, d.toInt()).toString()
                } catch (e: Exception) { /* not a real date */ }
            }
        }
        if (posted != null && TOMORROW.containsMatchIn(body)) {
            return nextSchoolDay(posted, schoolDays).toString()
        }
        return null
    }

    fun fromPost(post: Post, child: Child): List<ActionItem> {
        val body = post.title + "\n" + post.body
        val isHoliday = HOLIDAY.containsMatchIn(body)
        if (!ACTION.containsMatchIn(body) && !isHoliday) return emptyList()

        val today = LocalDate.now()
        val posted = postedDate(post)
        // Old posts create noise, not action: skip anything posted > 30 days ago.
        if (posted != null && posted.isBefore(today.minusDays(30))) return emptyList()

        val due = findDue(body, posted, schoolDays(child))
        // Already stale when it arrives -> not actionable, don't create it.
        if (due != null && LocalDate.parse(due).isBefore(today.minusDays(3))) return emptyList()

        val amount = AMOUNT.find(body)?.groupValues?.get(1)?.let { "₹$it" }
        val checklist = BULLET.findAll(body).take(8)
            .map { ChecklistEntry(it.groupValues[1].trim()) }.toMutableList()
        if (due == null && amount == null && checklist.isEmpty() && !isHoliday) return emptyList()

        val category = when {
            isHoliday -> "holiday"
            amount != null -> "fee"
            Regex("\\b(test|exam)\\b", RegexOption.IGNORE_CASE).containsMatchIn(body) -> "test"
            else -> "task"
        }
        val title = headline(body)
        return listOf(ActionItem(
            id = post.key + "#0",
            childEmail = post.childEmail,
            postKey = post.key,
            title = title,
            detail = snippet(post.body, title),
            dueDate = due, amount = amount, category = category,
            checklist = checklist))
    }

    private val DAY_TOKENS = buildMap {
        for ((i, key) in WEEKDAY_KEYS.withIndex()) {
            put(key, key)
            put(key.take(2), key)
            put(java.time.DayOfWeek.of(i + 1).getDisplayName(
                TextStyle.FULL, Locale.ENGLISH).lowercase(), key)
        }
    }

    /**
     * Best-effort parse of an OCR'd timetable image into day -> periods.
     * Grid OCR is imperfect; the result is meant to be reviewed in Settings.
     * Returns null when fewer than 3 day rows were recognised.
     */
    fun parseTimetableText(text: String): Map<String, List<String>>? {
        val result = linkedMapOf<String, MutableList<String>>()
        var current: String? = null
        for (raw in text.lines()) {
            val line = raw.trim()
            if (line.isEmpty()) continue
            val head = line.split(Regex("[:\\s]+"), limit = 2)
            val day = DAY_TOKENS[head[0].lowercase().trimEnd('.', ',')]
            if (day != null) {
                current = day
                result.getOrPut(day) { mutableListOf() }
                if (head.size > 1) {
                    result[day]!!.addAll(head[1].split(Regex("[,|]"))
                        .map { it.trim() }.filter { it.isNotEmpty() && it.length <= 16 })
                }
            } else if (current != null && line.length <= 16 &&
                       !line.contains(Regex("\\d{2}[:.]\\d{2}"))) {
                result[current]!!.add(line)
            }
        }
        val cleaned = result.filterValues { it.isNotEmpty() }
            .mapValues { it.value.take(9) }
        return if (cleaned.keys.size >= 3) cleaned else null
    }
}
